File: solutions/fsharp/clock/1/Clock.fs
module Clock

type Clock = { hours: int; minutes: int }

let create (hours: int) (minutes: int) : Clock =
    let total = (hours * 60 + minutes) % (24 * 60)

    match total with
    | t when t < 0 ->
        { hours = ((t + (24 * 60)) / 60) % 24
          minutes = (t + (24 * 60)) % 60 }
    | _ ->
        { hours = (total / 60) % 24
          minutes = total % 60 }

let add minutes clock =
    create clock.hours (clock.minutes + minutes)

let subtract minutes clock = add -minutes clock

let display clock =
    sprintf "%02d:%02d" clock.hours clock.minutes

File: solutions/go/clock/1/clock.go
// Package clock implements function for managing simple clocks.
package clock

import "fmt"

type Clock struct {
	hour   int
	minute int
}

// New returns a new clock given the hours and minutes.
func New(h, m int) Clock {
	total := (h*60 + m) % (24 * 60)

	if total < 0 {
		total += 24 * 60
	}

	return Clock{
		hour:   total / 60,
		minute: total % 60,
	}
}

// Add returns a new clock with m minutes added to the c clock.
func (c Clock) Add(m int) Clock {
	return New(c.hour, c.minute+m)
}

// Subtract returns a new clock with m minutes subtracted from the c clock.
func (c Clock) Subtract(m int) Clock {
	return c.Add(-m)
}

// String returns the string representation of the clock.
func (c Clock) String() string {
	return fmt.Sprintf("%02d:%02d", c.hour, c.minute)
}
